text/0000-ember-data-javascript-module-api.md

```diff
@@ -0,0 +1,130 @@
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Align Ember Data module API with [RFC #176](https://github.com/emberjs/rfcs/blob/master/text/0176-javascript-module-api.md).
+
+# Motivation
+
+This document proposes changes to the modules exported by Ember Data to make it consistent with the changes to the Ember module API proposed in [RFC #176](https://github.com/emberjs/rfcs/blob/master/text/0176-javascript-module-api.md).
+
+# Detailed design
+
+Ember Data will stick with 1 top level namespace of `@ember-data`. Under it there are 6 nested module 
+namespaces where exposed components could live. Someday these namespaces could be moved into their own 
+top level scoped package (for example `@ember-data/serializer` could become `@ember-data/serializer`) however,
+that is not being proposed at this time. 
+
+
+- `@ember-data/store` - Store related concers with storing records inside ember data's identity map.
+- `@ember-data/model` - Classes and utilities related to modeling data
+- `@ember-data/adapter` - Classes and utilities related to communicating with external backends
+- `@ember-data/serializer` - Classes for serializing and extracting data from outside Ember Data
+- `@ember-data/transform` - Classes for transforming individual values
+- `@ember-data/promise-proxies` - Classes for allowing promises to show up in ember templates.
+
+# How We Teach This
+
+What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing Ember patterns, or as a
+wholly new one?
+
+Would the acceptance of this proposal mean the Ember guides must be
+re-organized or altered? Does it change how Ember is taught to new users
+at any level?
+
+How should this feature be introduced and taught to existing Ember
+users?
+
+# Drawbacks
+
+Why should we *not* do this? Please consider the impact on teaching Ember,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+There are tradeoffs to choosing any path, please attempt to identify them here.
+
+# Alternatives
+
+The easiest alternative would be to continue using the current exports.
+Alternatively, we could use the current shims, which do not follow the RFC #176 heuristics as closely.
+
+# Unresolved questions
+
+## Should transforms live inside `ember-data/serializer`?
+
+Conceptually transforms are closly related to serializers,
+byt theyhave their own name and namespace in the Ember container.
+
+# Addenda
+
+## Addendum 1 - Table of Module Names and Exports by Global
+
+### `@ember-data/store`
+| Global                                | Module                                                                      |
+|---                                    | ---  |
+| `DS.Store` | `import Store from '@ember-data/store'` |
+| `DS.AdapterPopulatedRecordArray` | `import AdapterPopulatedRecordArray from '@ember-data/store/src/adapter-populated-record-array'` |
```

Comment:
`src` in the module import path is needed due to the new module unification semantics, which obviously makes it a little harder to build nice looking APIs based on import paths. IMHO we should first figure out how we could accomplish nice looking paths and then afterwards think about new Ember Data module APIs. Otherwise we might create more churn than needed.